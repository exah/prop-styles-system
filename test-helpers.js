/* eslint-env jest */

import { toArr, flatten, deepMerge } from '@exah/utils'
import renderer from 'react-test-renderer'

const THEME = {
  media: {
    D: '(min-width: 601px)',
    M: '(max-width: 600px)'
  }
}

const toJSON = (element) => renderer.create(element).toJSON()
const toStyles = (styles) => deepMerge(...flatten(toArr(styles)))
const wrapInMedia = (style) => ({ [`@media ${THEME.media.M}`]: style })

const testValue = ({
  fn,
  prop,
  cssProp,
  theme = THEME,
  values = [],
  autoValue
}) => () => {
  expect(toStyles(fn({ theme, [prop]: true }))).toEqual(autoValue != null ? { [cssProp]: autoValue } : {})
  expect(toStyles(fn({ theme, [prop]: 'auto' }))).toEqual(autoValue != null ? { [cssProp]: autoValue } : { [cssProp]: 'auto' })
  expect(toStyles(fn({ theme, [prop]: null }))).toEqual({})

  values.forEach((val) => {
    expect(toStyles(fn({ theme, [prop]: val }))).toEqual({ [cssProp]: val })
    expect(toStyles(fn({ theme, [prop]: val }))).toEqual({ [cssProp]: val })
    expect(toStyles(fn({ theme, [prop]: { M: val } }))).toEqual(wrapInMedia({ [cssProp]: val }))
  })
}
export {
  toJSON,
  THEME as theme,
  toStyles,
  wrapInMedia,
  testValue
}
