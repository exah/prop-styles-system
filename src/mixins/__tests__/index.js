/* eslint-env jest */

import 'jest-styled-components'
import styled from 'styled-components'
import { createElement as h } from 'react'
import { mq, prop, themePath } from '../..'
import { toJSON } from '../../../test-helpers'

const theme = {
  media: {
    sm: '(max-width: 600px)'
  },
  space: [ 0, 8, 16, 32, 64 ],
  color: {
    blue: '#0000FF'
  },
  palette: {
    default: {
      bg: '#ffffff',
      fg: '#000000'
    },
    inverted: {
      bg: '#000000',
      fg: '#ffffff'
    }
  }
}

describe('mixin', () => {
  const Box = styled('div')`
    background-color: ${prop('bg', 'blue')};
    color: ${prop('fg', themePath('color.blue'))};
    border-color: ${prop('borderColor')};

    @media ${mq('sm')} {
      background-color: black;
    }
  `

  test('prop', () => {
    const tree = toJSON(h(Box, { theme, bg: 'red', fg: 'red' }))
    expect(tree).toHaveStyleRule('background-color', 'red')
    expect(tree).toHaveStyleRule('color', 'red')
    expect(tree).not.toHaveStyleRule('border-color')
  })

  test('mq', () => {
    const tree = toJSON(h(Box, { theme }))
    expect(tree).toHaveStyleRule('background-color', 'black', { media: theme.media.sm })
  })

  test('fallbacks', () => {
    const tree = toJSON(h(Box, { theme }))
    expect(tree).toHaveStyleRule('background-color', 'blue')
    expect(tree).toHaveStyleRule('color', '#0000FF')
    expect(tree).toMatchSnapshot()
  })
})
