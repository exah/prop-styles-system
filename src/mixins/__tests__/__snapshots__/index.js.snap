// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mixin fallbacks 1`] = `
.c0 {
  background-color: blue;
  color: #0000FF;
  width: 100%;
}

@media (max-width:600px) {
  .c0 {
    background-color: black;
  }
}

<div
  className="c0"
/>
`;
